### PR TITLE
TELCODOCS-1109

### PR DESCRIPTION
**Pull request by StephenJamesSmith**
Create KMMO doc:  Hub & Spoke

Jira: https://issues.redhat.com/browse/TELCODOCS-1109

Version(s): openshift-4.15, openshift-4.14, openshift-4.13.z, openshift-4.12.z

Link to docs preview: https://67822--docspreview.netlify.app/openshift-enterprise/latest/hardware_enablement/kmm-kernel-module-management#kmm-hub-hub-and-spoke_kernel-module-management-operator

SME review: @bthurber 
QE review: @cdvultur

### CODE REVIEW

**Comment by StephenJamesSmith** [comment]
@bthurber @cdvultur Please review. Thx!

**Comment by bthurber** [comment]
@qbarrand lgtm.  Do you mind double checking?

File: modules/kmm-hub-hub-and-spoke.adoc [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Module included in the following assemblies:
+//
+// * hardware_enablement/kmm-kernel-module-management.adoc
+
+:_content-type: CONCEPT
+[id="kmm-hub-hub-and-spoke_{context}"]
+= KMM hub & spoke
+
+In Kernel Module Management (KMM) hub & spoke scenarios, many spoke clusters are connected to a central, powerful hub cluster. Open Cluster Management (OCM) is the reference open source implementation of this pattern, and KMM depends on it to operate in those scenarios.
```
Comment (cdvultur) [inline modules/kmm-hub-hub-and-spoke.adoc:9]: Red Hat documentation should point to RHACM ( Red Hat Advanced Cluster Management and not to the Upstream counterpart. 
Even the Midstream document points to that: https://openshift-kmm.netlify.app/documentation/hub_spoke/

File: modules/kmm-hub-installing-kmm-hub-kubectl.adoc [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+//
+// * hardware_enablement/kmm-kernel-module-management.adoc
+
+:_content-type: PROCEDURE
+[id="kmm-hub-installing-kmm-hub-kubectl_{context}"]
+= Installing KMM-Hub using `kubectl`
+
+Use this procedure to install KMM-Hub using `kubectl`.
+
+.Procedure
```
Comment (cdvultur) [inline modules/kmm-hub-installing-kmm-hub-kubectl.adoc:11]: This is the procedure for U/S .

CertManager is required for Upstream version of KMM-HUB

File: modules/kmm-hub-installing-kmm-hub-olm.adoc [hunk 3]
```diff
@@ -0,0 +1,18 @@
+// Module included in the following assemblies:
+//
+// * hardware_enablement/kmm-kernel-module-management.adoc
+
+:_content-type: PROCEDURE
+[id="kmm-hub-installing-kmm-hub-olm_{context}"]
+= Installing KMM-Hub using the Operator Lifecycle Manager
+
+Use this procedure to install KMM-Hub using the Operator Lifecycle Manager (OLM).
```
Comment (cdvultur) [inline modules/kmm-hub-installing-kmm-hub-olm.adoc:9]: Same thing here. In the Openshift docs we should reference installing KMM-HUB from redhat-operators catalog via UI or via example from M/S docs: 
https://openshift-kmm.netlify.app/documentation/hub_spoke/#on-the-hub

**Review by cdvultur — CHANGES_REQUESTED** [review]
I think the work for this PR was based on U/S docs of KMM: https://kmm.sigs.k8s.io/documentation/hub_spoke/

When it should have been done based on M/S : https://openshift-kmm.netlify.app/documentation/hub_spoke/

For the official OpenShift documentation I suggest using the M/S repository

**Comment by StephenJamesSmith** [comment]
@qbarrand Could you double check this?

**Comment by qbarrand** [comment]
@StephenJamesSmith I agree with @cdvultur's comments: our docs should point to RHACM, and use the midstream docs as base - not upstream.

File: modules/kmm-hub-installing-kmm-hub-kubectl.adoc (hunk 2 above)
Comment (StephenJamesSmith) [reply modules/kmm-hub-installing-kmm-hub-kubectl.adoc:11]: This section has been removed.

File: modules/kmm-hub-hub-and-spoke.adoc (hunk 1 above)
Comment (StephenJamesSmith) [reply modules/kmm-hub-hub-and-spoke.adoc]: Rewrote these sections to refer to RHACM M/S content.

File: modules/kmm-hub-installing-kmm-hub-olm.adoc (hunk 3 above)
Comment (StephenJamesSmith) [reply modules/kmm-hub-installing-kmm-hub-olm.adoc]: Rewrote these sections to refer to RHACM M/S content.

**Comment by StephenJamesSmith** [comment]
@qbarrand @cdvultur Please review content rewritten for RHACM M/S.

File: modules/kmm-hub-hub-and-spoke.adoc [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// Module included in the following assemblies:
+//
+// * hardware_enablement/kmm-kernel-module-management.adoc
+
+:_content-type: CONCEPT
+[id="kmm-hub-hub-and-spoke_{context}"]
+= KMM Hub & Spoke
+
+In Kernel Module Management (KMM) Hub & Spoke scenarios, many spoke clusters are connected to a central, powerful hub cluster. KMM depends on Red Hat Advanced Cluster Management (RHACM) to operate in  Hub & Spoke environments.
```
Comment (qbarrand) [inline modules/kmm-hub-hub-and-spoke.adoc:9]:
```suggestion
In Hub & Spoke scenarios, many spoke clusters are connected to a central, powerful hub cluster. KMM depends on Red Hat Advanced Cluster Management (RHACM) to operate in  Hub & Spoke environments.
```
Not a KMM-specific concept.

File: modules/kmm-hub-hub-and-spoke.adoc [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+KMM is compatible with hub & spoke environments through decoupling KMM features. A ``ManagedClusterModule`` Custom Resource Definition (CRD) is provided to wrap the existing ``Module`` CRD and extend it to select Spoke clusters. Also provided is KMM-Hub, a new standalone controller that builds images and signs modules on the hub cluster.
+
+In hub & spoke setups, spokes are focused, resource-constrained clusters that are centrally managed by a hub cluster. Spokes run the single-cluster edition of KMM, with those resource-intensive features disabled.To adapt KMM to this environment, it is desirable to reduce the workload running on the spokes to the minimum, while the hub takes care of the expensive tasks.
```
Comment (qbarrand) [inline modules/kmm-hub-hub-and-spoke.adoc:13]:
```suggestion
In hub & spoke setups, spokes are focused, resource-constrained clusters that are centrally managed by a hub cluster. Spokes run the single-cluster edition of KMM, with those resource-intensive features disabled. To adapt KMM to this environment, it is desirable to reduce the workload running on the spokes to the minimum, while the hub takes care of the expensive tasks.
```

File: modules/kmm-hub-installing-kmm-hub-kubectl.adoc
Comment (qbarrand) [inline modules/kmm-hub-installing-kmm-hub-kubectl.adoc:1]: This still references upstream. Please remove that file entirely as we only want to support OLM bundles as the installation medium.

File: modules/kmm-hub-installing-kmm-hub-cli.adoc
Comment (qbarrand) [inline modules/kmm-hub-installing-kmm-hub-cli.adoc:1]: Please remove that file entirely as we only want to support OLM bundles as the installation medium.

File: modules/kmm-hub-hub-and-spoke.adoc
Comment (qbarrand) [inline modules/kmm-hub-hub-and-spoke.adoc:1]: Shall we use capitalized Hub and Spoke words to be consistent in all files?

File: modules/kmm-hub-hub-and-spoke.adoc (hunk 5 above)
Comment (StephenJamesSmith) [reply modules/kmm-hub-hub-and-spoke.adoc:13]: Changed.

File: modules/kmm-hub-hub-and-spoke.adoc (hunk 4 above)
Comment (StephenJamesSmith) [reply modules/kmm-hub-hub-and-spoke.adoc:9]: Changed.

File: modules/kmm-hub-installing-kmm-hub-cli.adoc
Comment (StephenJamesSmith) [reply modules/kmm-hub-installing-kmm-hub-cli.adoc:1]: @qbarrand The contents of this file document the `oc apply` command from the M/S. Please confirm you want this removed.

File: modules/kmm-hub-hub-and-spoke.adoc
Comment (qbarrand) [reply modules/kmm-hub-hub-and-spoke.adoc:1]: Looks like other bits of the OpenShift documentation indeed use lower "hub" and "spoke". Let us make it lowercase everywhere then. I'll note that [that page](https://docs.openshift.com/container-platform/4.14/scalability_and_performance/ztp_far_edge/ztp-deploying-far-edge-clusters-at-scale.html#about-ztp_ztp-deploying-far-edge-clusters-at-scale) refers to "hub-and-spoke architecture", but I am not sure if we indeed want to use hyphens - I'll leave that decision to you.

File: modules/kmm-hub-installing-kmm-hub-cli.adoc
Comment (qbarrand) [reply modules/kmm-hub-installing-kmm-hub-cli.adoc:1]: Please remove. We want our customers to use bundles only.

File: modules/kmm-hub-hub-and-spoke.adoc
Comment (StephenJamesSmith) [reply modules/kmm-hub-hub-and-spoke.adoc:1]: @qbarrand Let's use "hub and spoke"  for now and see what happens at Merge review. Please review and Ack latest changes. Thx!

**Comment by StephenJamesSmith** [comment]
@cdvultur Please review and Ack (/lgtm) latest changes.

File: modules/kmm-hub-kmm-hub.adoc [hunk 6]
```diff
@@ -0,0 +1,16 @@
+// Module included in the following assemblies:
+//
+// * hardware_enablement/kmm-kernel-module-management.adoc
+
+:_content-type: CONCEPT
+[id="kmm-hub-kmm-hub_{context}"]
+= KMM-Hub
+
+The KMM project provides KMM-Hub, an edition of KMM dedicated to hub clusters. KMM-Hub is aware of all kernel versions running on the spokes and determines the nodes on the cluster that should receive a kernel module.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-hub-kmm-hub.adoc:9]: Haha. Changed to "KMM-Hub monitors all kernel versions running on the spokes and determines the nodes on the cluster that should receive a kernel module."

File: modules/kmm-hub-installing-kmm-hub-olm.adoc [hunk 7]
```diff
@@ -0,0 +1,9 @@
+// Module included in the following assemblies:
+//
+// * hardware_enablement/kmm-kernel-module-management.adoc
+
+:_content-type: PROCEDURE
+[id="kmm-hub-installing-kmm-hub-olm_{context}"]
+= Installing KMM-Hub using the Operator Lifecycle Manager
+
+The preferred way to install KMM-Hub is to use the *Operators* section of the OpenShift console.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-hub-installing-kmm-hub-olm.adoc:9]: Changed.

File: modules/kmm-hub-hub-and-spoke.adoc [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+[id="kmm-hub-hub-and-spoke_{context}"]
+= KMM hub and spoke
+
+In hub and spoke scenarios, many spoke clusters are connected to a central, powerful hub cluster. KMM depends on Red Hat Advanced Cluster Management (RHACM) to operate in  Hub & Spoke environments.
+
+KMM is compatible with hub and spoke environments through decoupling KMM features. A ``ManagedClusterModule`` Custom Resource Definition (CRD) is provided to wrap the existing ``Module`` CRD and extend it to select Spoke clusters. Also provided is KMM-Hub, a new standalone controller that builds images and signs modules on the hub cluster.
+
+In hub and spoke setups, spokes are focused, resource-constrained clusters that are centrally managed by a hub cluster. Spokes run the single-cluster edition of KMM, with those resource-intensive features disabled. To adapt KMM to this environment, you should reduce the workload running on the spokes to the minimum, while the hub takes care of the expensive tasks.
+
+Building kernel module images, as well as signing the ``.ko`` files, should run on the hub, while the scheduling of the Module Loader and Device Plugin ``DaemonSets`` can only happen on the spokes.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-hub-hub-and-spoke.adoc:15]: Made these changes.

File: hardware_enablement/kmm-kernel-module-management.adoc [hunk 9]
```diff
@@ -83,6 +83,29 @@ include::modules/kmm-building-and-signing-a-moduleloader-container-image.adoc[le
 .Additional resources
 For information on creating a service account, see link:https://docs.openshift.com/container-platform/4.12/authentication/understanding-and-creating-service-accounts.html#service-accounts-managing_understanding-service-accounts[Creating service accounts].
 
+// Added for TELCODOCS-1109
+include::modules/kmm-hub-hub-and-spoke.adoc[leveloffset=+1]
+[role="_additional-resources"]
+.Additional resources
+* link:https://www.redhat.com/en/technologies/management/advanced-cluster-management[Red Hat Advanced Cluster Management (RHACM)]
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply hardware_enablement/kmm-kernel-module-management.adoc:90]: Huh - Never knew that. Added {nbsp}.

File: modules/kmm-hub-hub-and-spoke.adoc [hunk 10]
```diff
@@ -0,0 +1,15 @@
+// Module included in the following assemblies:
+//
+// * hardware_enablement/kmm-kernel-module-management.adoc
+
+:_content-type: CONCEPT
+[id="kmm-hub-hub-and-spoke_{context}"]
+= KMM hub and spoke
+
+In hub and spoke scenarios, many spoke clusters are connected to a central, powerful hub cluster. KMM depends on Red Hat Advanced Cluster Management (RHACM) to operate in  Hub & Spoke environments.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-hub-hub-and-spoke.adoc:9]: Added {nbsp}.

**Comment by nalhadef** [comment]
The nbsp in Red Hat was only recently added to the RHSSG.

I've added a lgtm label. Thanks!!

File: modules/kmm-hub-running-kmm-on-the-spoke.adoc [hunk 11]
```diff
@@ -0,0 +1,117 @@
+// Module included in the following assemblies:
+//
+// * hardware_enablement/kmm-kernel-module-management.adoc
+
+:_content-type: PROCEDURE
+[id="kmm-hub-running-kmm-on-the-spoke_{context}"]
+= Running KMM on the spoke
+
+After installing KMM on the spoke, no further action is required. Create `ManagedClusterModule` from the hub to deploy kernel modules on spoke clusters.
```
Comment (mburke5678) [inline modules/kmm-hub-running-kmm-on-the-spoke.adoc:9]:
```suggestion
After installing KMM on the spoke, no further action is required. Create a `ManagedClusterModule` object from the hub to deploy kernel modules on spoke clusters.
```
or 
```suggestion
After installing KMM on the spoke, no further action is required. Create a managed cluster module from the hub to deploy kernel modules on spoke clusters.
```

File: modules/kmm-hub-using-the-managedclustermodule.adoc [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+  moduleSpec: <1>
+    selector: <2>
+      node-wants-my-mcm: 'true'
+
+  spokeNamespace: <some-namespace> <3>
+
+  selector:
+    wants-my-mcm: 'true'  # Selects ManagedCluster objects
+----
+<1> `moduleSpec`: Contains `moduleLoader` and `devicePlugin` sections, just like in a `Module` resource.
```
Comment (mburke5678) [inline modules/kmm-hub-using-the-managedclustermodule.adoc:29]:
```suggestion
<1> Contains the `moduleLoader` and `devicePlugin` sections, similar to a `Module` resource.
```
Do we want to show the `devicePlugin`? 
What about the spaces (Line 23 and 25? Should they be removed?

File: modules/kmm-hub-using-the-managedclustermodule.adoc [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+<2> Selects nodes within the `ManagedCluster`.
```
Comment (mburke5678) [inline modules/kmm-hub-using-the-managedclustermodule.adoc:31]:
```suggestion
<2> Selects nodes within the managed cluster.
```

**Comment by mburke5678** [comment]
@StephenJamesSmith A few nits I was hoping you would consider before merging.

File: modules/kmm-hub-hub-and-spoke.adoc [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+//
+// * hardware_enablement/kmm-kernel-module-management.adoc
+
+:_content-type: CONCEPT
+[id="kmm-hub-hub-and-spoke_{context}"]
+= KMM hub and spoke
+
+In hub and spoke scenarios, many spoke clusters are connected to a central, powerful hub cluster. Kernel Module Management (KMM) depends on Red{nbsp}Hat Advanced Cluster Management (RHACM) to operate in hub and spoke environments.
+
+KMM is compatible with hub and spoke environments through decoupling KMM features. A ``ManagedClusterModule`` Custom Resource Definition (CRD) is provided to wrap the existing ``Module`` CRD and extend it to select Spoke clusters. Also provided is KMM-Hub, a new standalone controller that builds images and signs modules on the hub cluster.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-hub-hub-and-spoke.adoc:11]: That happens when I bring stuff over from another doc that uses another markdown language. I typically try to catch them. I'll change them.

File: modules/kmm-hub-kmm-hub.adoc [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+//
+// * hardware_enablement/kmm-kernel-module-management.adoc
+
+:_content-type: CONCEPT
+[id="kmm-hub-kmm-hub_{context}"]
+= KMM-Hub
+
+The KMM project provides KMM-Hub, an edition of KMM dedicated to hub clusters. KMM-Hub monitors all kernel versions running on the spokes and determines the nodes on the cluster that should receive a kernel module.
+
+KMM-Hub runs all compute-intensive tasks such as image builds and kmod signing, and prepares the trimmed-down `Module` to be transferred to the spokes through RHACM.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-hub-kmm-hub.adoc:11]: Changed.

File: modules/kmm-hub-running-kmm-on-the-spoke.adoc (hunk 11 above)
Comment (StephenJamesSmith) [reply modules/kmm-hub-running-kmm-on-the-spoke.adoc:9]: Changed.

File: modules/kmm-hub-using-the-managedclustermodule.adoc [hunk 16]
```diff
@@ -0,0 +1,36 @@
+// Module included in the following assemblies:
+//
+// * hardware_enablement/kmm-kernel-module-management.adoc
+
+:_content-type: PROCEDURE
+[id="kmm-hub-using-the-managedclustermodule_{context}"]
+= Using the `ManagedClusterModule` CRD
+
+Use the `ManagedClusterModule` Custom Resource Definition (CRD) to configure the deployment of kernel modules on spoke clusters.
+This CRD is cluster-scoped, wraps a ``Module`` spec and adds the following additional fields:
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-hub-using-the-managedclustermodule.adoc:10]: Changed.

File: modules/kmm-hub-using-the-managedclustermodule.adoc (hunk 12 above)
Comment (StephenJamesSmith) [reply modules/kmm-hub-using-the-managedclustermodule.adoc:29]: Changed "just like in...". No other changes.

File: modules/kmm-hub-using-the-managedclustermodule.adoc [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+  # No namespace, because this resource is cluster-scoped.
+spec:
+  moduleSpec: <1>
+    selector: <2>
+      node-wants-my-mcm: 'true'
+
+  spokeNamespace: <some-namespace> <3>
+
+  selector:
+    wants-my-mcm: 'true'  # Selects ManagedCluster objects
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-hub-using-the-managedclustermodule.adoc:27]: Changed to a call-out.

File: modules/kmm-hub-using-the-managedclustermodule.adoc (hunk 13 above)
Comment (StephenJamesSmith) [reply modules/kmm-hub-using-the-managedclustermodule.adoc:31]: No change.

**Comment by StephenJamesSmith** [comment]
@mburke5678 Made most suggestions. Please Merge. Thx!

**Comment by StephenJamesSmith** [comment]
@mburke5678 Please merge https://github.com/openshift/openshift-docs/pull/68887

**Pull request merged**